File: backend/src/main/java/com/opstrack/inspection/InspectionService.java
package com.opstrack.inspection;

import com.opstrack.maintenance.MaintenanceStatus;
import com.opstrack.maintenance.MaintenanceTask;
import com.opstrack.maintenance.MaintenanceTaskRepository;
import com.opstrack.technician.Technician;
import com.opstrack.technician.TechnicianRepository;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class InspectionService {

    private final InspectionRepository inspectionRepository;
    private final MaintenanceTaskRepository maintenanceTaskRepository;
    private final TechnicianRepository technicianRepository;

    public InspectionService(
            InspectionRepository inspectionRepository,
            MaintenanceTaskRepository maintenanceTaskRepository,
            TechnicianRepository technicianRepository
    ) {
        this.inspectionRepository = inspectionRepository;
        this.maintenanceTaskRepository = maintenanceTaskRepository;
        this.technicianRepository = technicianRepository;
    }

    public Inspection createInspection(
            Long maintenanceTaskId,
            Long inspectorId,
            InspectionStatus status,
            String comments
    ) {
        MaintenanceTask task =
                maintenanceTaskRepository.findById(maintenanceTaskId)
                        .orElseThrow(() -> new IllegalArgumentException(
                                "Maintenance task not found with id: "
                                        + maintenanceTaskId
                        ));

        Technician inspector =
                technicianRepository.findById(inspectorId)
                        .orElseThrow(() -> new IllegalArgumentException(
                                "Inspector not found with id: "
                                        + inspectorId
                        ));

        Inspection inspection = new Inspection(
                status,
                comments,
                task,
                inspector
        );

        if (status == InspectionStatus.APPROVED) {
            task.setStatus(MaintenanceStatus.COMPLETED);
        }

        if (status == InspectionStatus.REJECTED) {
            task.setStatus(MaintenanceStatus.IN_PROGRESS);
        }

        maintenanceTaskRepository.save(task);

        return inspectionRepository.save(inspection);
    }

    public List<Inspection> getAllInspections() {
        return inspectionRepository.findAll();
    }

    public List<Inspection> getInspectionsByMaintenanceTaskId(
            Long maintenanceTaskId
    ) {
        return inspectionRepository.findByMaintenanceTaskId(
                maintenanceTaskId
        );
    }
}

File: frontend/src/app/inspection/inspection.service.ts
import { API_BASE_URL } from '../api.config';
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';

import {
  MaintenanceTask
} from '../maintenance/maintenance-task.service';

import {
  Technician
} from '../technician/technician.service';

export type InspectionStatus =
  | 'PENDING'
  | 'APPROVED'
  | 'REJECTED';

export interface Inspection {
  id: number;
  status: InspectionStatus;
  comments: string;
  inspectedAt: string;
  maintenanceTask: MaintenanceTask;
  inspector: Technician;
}

@Injectable({
  providedIn: 'root'
})
export class InspectionService {

  private readonly apiUrl =
    `${API_BASE_URL}/api/inspections`;

  constructor(
    private readonly http: HttpClient
  ) {}

  getAllInspections(): Observable<Inspection[]> {

    return this.http.get<Inspection[]>(
      this.apiUrl
    );
  }

  getInspectionsByMaintenanceTaskId(
    maintenanceTaskId: number
  ): Observable<Inspection[]> {

    return this.http.get<Inspection[]>(
      `${this.apiUrl}/maintenance-task/${maintenanceTaskId}`
    );
  }

  createInspection(
    maintenanceTaskId: number,
    inspectorId: number,
    status: InspectionStatus,
    comments: string
  ): Observable<Inspection> {

    return this.http.post<Inspection>(
      this.apiUrl,
      null,
      {
        params: {
          maintenanceTaskId,
          inspectorId,
          status,
          comments
        }
      }
    );
  }
}
